Implement mulled containers v2.

jmchilton opened the pull request:
For a given set of packages and versions, the v1 containers had names and versions like:

```
mulled-v1-<hash_of_packages_and_versions>:<build_number>
```

Here the quay.io repository has a unique name per combination of packages and versions - and the repository tag (the part after the :) is solely used to specify a build number (a mechanism allowing us to increment and rebuild containers with same recipes if they are broken in some way - mirroring closely the conda build number).

The version of the v2 hash looks like:

```
mulled-v2-<hash_of_packages>:<hash_of_versions>-<build_number>
```

Here a repository is only created for each combination of packages - different combinations of versions are represented as different tags within that repository.

This commit preserves backward compatibility - but since we are aware of no biocontainers published in the wild it would probably be a okay idea to eliminate the backward compatability once we ready to merge into Galaxy - at least when looking for published containers.

Ping @bgruening - I still need to port this over to Galaxy and test it - but based on the PR description do you think this is a good course forward? I also want to special case a list of requirements where none of them specify a version - in that case we should probably not generate a version hash and simply rely on build numbers.

bgruening commented on galaxy/tools/deps/container_resolvers/mulled.py:
```diff
@@ -85,14 +94,36 @@ def cached_container_description(targets, namespace):
-                image = cached_image
-                break
+            if version_hash is None:
+                # Match on package and version hash...
```
The comments are mixed up here and below I think.

bgruening commented on galaxy/tools/deps/mulled/util.py:
```diff
@@ -123,6 +129,31 @@ def image_name(targets, image_build=None, name_override=None):
         return "mulled-v1-%s%s" % (m.hexdigest(), suffix)
 
 
+def v2_image_name(targets, image_build=None, name_override=None):
```
This function really needs a doc test, so that people can see what is hashed into what and can reimplement the hashing.

jmchilton replied:
Sure - good idea.

bgruening commented:
Description ready exactly like what we discussed! :+1: 
I don't think we need to preserve backward backward compatibility. 

I wonder what we do with plain multi-conda environments in Galaxy. Do they stay as they are or do we want to unify the names? Nate will probably hat us ;)

Thanks a bunch! Sleep well!

jmchilton commented on galaxy/tools/deps/container_resolvers/mulled.py:
```diff
@@ -47,7 +50,13 @@ def output_line_to_image(line):
+            if version and "-" in version:
+                version_hash, build = version.split("-", 1)
+            else:
+                version_hash = None
```
Yup - I hadn't tested that no versions case - I'll make another pass at this and fix that.

jmchilton commented:
> I wonder what we do with plain multi-conda environments in Galaxy. Do they stay as they are or do we want to unify the names?

 I don't see a compelling reason to either force rebuilds for v2 or prohibit v2 naming. I guess in both cases we just provide both options and default Conda to v1 and containers to v2.

bgruening commented:
> I don't see a compelling reason to either force rebuilds for v2 or prohibit v2 naming. I guess in both cases we just provide both options and default Conda to v1 and containers to v2.

At some point it would be nice to align both so people can map conda envs to Docker/Singularity containers.

jmchilton commented:
> At some point it would be nice to align both so people can map conda envs to Docker/Singularity containers.

Agreed - maybe just for new installs though? I don't want to break the tooling and re-computability of people already using Conda.

bgruening commented:
Oh for sure, the old ones needs to stay, but maybe we can switch the news ones at some point.

jmchilton commented:
Okay - I have done a bunch of fixes and rebased them into that original commit. I have tried this within Galaxy and I have tested both a normal container and a container with two versionless tools - I think this is ready to go to now. I added a TODO in there about sorting based on build number - but I think that was an issue prior to these changes.

I'll open a matching Galaxy PR shortly.

bgruening commented:
Great work @jmchilton! Sorry for not having this idea in the first place. It feels more correct!
